File: target_includes/ios_bridges/Button.swift
//
//  Button.swift
//  MobileApplicationsSwiftAufgabe1
//
//  Created by Tobias Markus on 21.04.15.
//  Copyright (c) 2015 Tobias Markus. All rights reserved.
//

import Foundation
import UIKit

/**
* Class that represents a button
* @author tobiasmarkus
*/
public class Button {
    
    private var innerButton: UIButton;
    private var x: CGFloat = 0, y: CGFloat = 0;
    /**
    * Specifies the place where the textfield goes.
    */
    private var parentUIContext: UIViewController;
    
    /**
    * Specifies the place where the handlers are defined
    */
    private var parentEventContext: UIResponder;
    
    /**
    * Public constructor of class "Button"
    */
    public init(context: UIViewController!, eventContext: UIResponder!)
    {
        // Set parent context:
        self.parentUIContext = context;
        self.parentEventContext = eventContext;
        
        // Initialize button
        self.innerButton = UIButton(frame: CGRect(x: 0, y: 0, width: 100, height: 30));
        
        // Set the default look for this button
        self.innerButton.backgroundColor = UIColor.whiteColor();
        self.innerButton.setTitleColor(UIColor.blackColor(), forState: UIControlState.Normal);
        self.innerButton.setTitleColor(UIColor.blueColor(), forState: UIControlState.Selected);
        self.innerButton.layer.borderWidth = 1;
        self.innerButton.layer.cornerRadius = 5;
        if let titleLabel = self.innerButton.titleLabel
        {
            titleLabel.font = UIFont(name: "Arial", size: 12);
        }
    }
    
    /**
    * Returns the label for this button
    * @return String The label for this element
    */
    public func getLabel() -> String
    {
        if let titleLabel = self.innerButton.titleLabel
        {
            if let text = titleLabel.text
            {
                return text;
            }
        }
        
        return "";
    }
    
    /**
    * Sets the label for this button
    * @param String The label to set
    */
    public func setLabel(label: String)
    {
        self.innerButton.setTitle(label, forState: UIControlState.Normal);
    }
    
    /**
    * Sets the size of this element
    * @param width The width of this element in percent
    * @param height The height of this element in percent
    */
    public func setSize(width: CGFloat, height: CGFloat)
    {
        var screenWidth = (self.parentUIContext as! ApplicationView).getWidth();
        var screenHeight = (self.parentUIContext as! ApplicationView).getHeight();

        var newWidth = (screenWidth / 100) * width;
        var newHeight = (screenHeight / 100) * height;
        
        self.innerButton.frame = CGRectMake(self.x, self.y, newWidth, newHeight);
    }
    
    /**
    * Sets the position of this element
    * @param x The x position of this element
    * @param y The y position of this element
    */
    public func setPosition(x: CGFloat, y: CGFloat)
    {
        var screenWidth = (self.parentUIContext as! ApplicationView).getWidth();
        var screenHeight = (self.parentUIContext as! ApplicationView).getHeight();

        var newX = (screenWidth / 100) * x;
        var newY = (screenHeight / 100) * y;

        self.x = newX;
        self.y = newY;
        self.innerButton.frame = CGRectMake(newX, newY, self.innerButton.frame.width, self.innerButton.frame.height);
    }
    
    /**
    * Gets the wrapped element
    */
    public func getWrappedElement() -> UIButton
    {
        return innerButton;
    }
    
    /**
     * Adds this button to the specified application view
     */
    public func addToApplicationView()
    {
        parentUIContext.view.addSubview(getWrappedElement());
    }
    
    /**
     * Adds an onClick listener to this button
     * @param listenerName Function name of the listener
     */
    public func addOnClickListener(methodName: String) {
        // add a target
        innerButton.addTarget(parentEventContext, action: Selector(methodName + ":"), forControlEvents: UIControlEvents.TouchUpInside);
    }
}

File: target_includes/android_bridges/Button.java
package de.fhflensburg.tobiasmarkus.androidBridge;

import android.content.Context;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup.LayoutParams;

import java.lang.reflect.InvocationTargetException;

/**
 * Class that represents a button
 * @author tobiasmarkus
 *
 */
public class Button {
	
	/**
	 * Private button instance
	 */
	private android.widget.Button button;
	
	/**
	 * Private layout params instance
	 */
	private LayoutParams layoutParams = null;
	
	/**
	 * Parent context (activity?)
	 */
	private Context parentContext = null;
	
	/**
	 * Parent view
	 */
	private ApplicationView parentView;

	/**
	 * Public constructor of class Button
	 */
	public Button(ApplicationView parentView)
	{
		this.parentView = parentView;
		parentContext = parentView.getParentContext();
		button = new android.widget.Button(parentContext);
		button.setTextSize(12);
	}
	
	/**
	 * Returns the label of this button
	 * @return String The label of this button
	 */
	public String getLabel()
	{
		return button.getText().toString();
	}
	
	/**
	 * Sets the label for this button
	 * @param label The label for this button
	 */
	public void setLabel(String label)
	{
		button.setText(label);
	}
	
	/**
	 * Sets the size of this element
	 * @param width The width of this element in percent
	 * @param height The height of this element in percent
	 */
	public void setSize(float width, float height)
	{
		int newWidth = (int)((parentView.getWidth() / 100) * width);
		int newHeight = (int)(parentView.getHeight() / 100 * height);
		button.setWidth(newWidth);
		button.setHeight(newHeight);
		layoutParams = new LayoutParams(newWidth, newHeight);
		button.setLayoutParams(layoutParams);
	}
	
	/**
	 * Sets the position of this element
	 * @param x The x position of this element
	 * @param y The y position of this element
	 */
	public void setPosition(float x, float y)
	{
		int newX = (int)((parentView.getWidth() / 100) * x);
		int newY = (int)(parentView.getHeight() / 100 * y);
		
		button.setX(newX);
		button.setY(newY);
	}
	
	/**
	 * Gets the wrapped element
	 */
	public android.widget.Button getWrappedElement() {
		return button;
	}
	
	/**
	 * Adds this button to the specified activity.
	 * @param activity The activity to add the button to
	 */
	public void addToApplicationView() {
		((android.app.Activity)parentContext).addContentView(this.getWrappedElement(), this.getWrappedElement().getLayoutParams());
	}
	
	/**
	 * Adds an onClick listener to this element
	 * @param listener The listener to add to this element.
	 */
	public void addOnClickListener(final String methodName) {
		final Button _button = this;
		button.setOnClickListener(new OnClickListener() {
			
			@Override
			public void onClick(View v) {
				java.lang.reflect.Method method = null;
				try {
				  method = parentContext.getClass().getMethod(methodName, Button.class);
				} catch (SecurityException e) {
				  // ...
				} catch (NoSuchMethodException e) {
				  // ...
					e.printStackTrace();
				}
				if(method == null)
					return;

				try {
					method.invoke(parentContext, _button);
				} catch (IllegalAccessException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				} catch (IllegalArgumentException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				} catch (InvocationTargetException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		});
	}
}